table/tables/partition.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tables
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/mock"
+	log "github.com/sirupsen/logrus"
+)
+
+// Table, Partition, PartitionTable both implements the table.Table interface.
+var _ table.Table = &Table{}
+var _ table.Table = &Partition{}
+var _ table.Table = &PartitionTable{}
+
+// PartitionTable implements the table.PartitionTable interface.
+var _ table.PartitionTable = &PartitionTable{}
+
+// Partition implements the table.Table interface.
+type Partition struct {
+	tableCommon
+}
+
+// PartitionTable implements the table.PartitionTable interface.
+type PartitionTable struct {
+	Table
+	partitionExpr *PartitionExpr
+}
+
+// PartitionExpr is the partition definition expressions.
+// There are two expressions exist, because Locate use binary search, which requires:
+// Given a compare function, for any partition range i, if cmp[i] > 0, then cmp[i+1] > 0.
+// While partition prune must use the accurate range to do prunning.
+// partition by range (x)
+//   (partition
+//      p1 values less than (y1)
+//      p2 values less than (y2)
+//      p3 values less than (y3))
+// PartitionPrune: (x < y1); (y1 <= x < y2); (y2 <= x < y3)
+// Locate: (x < y1); (x < y2); (x < y3)
+type PartitionExpr struct {
+	PartitionPrune []expression.Expression
```

Comment:
`Ranges` is better than `PartitionPrune`

Reply:
The naming convention is following their usage here.
`PartitionPrune` is used by logical plan optimize rule, `Locate` is used by function `locatePartition`.

All those code are not new, they are just moved.

Reply:
I can not agree with this:
>The naming convention is following their usage here.

The name should be what it is, rather than what it is used for.

Reply:
Should I name the Locate field? Any suggestion?